adapters/repos/db/queue/queue.go

```diff
@@ -0,0 +1,747 @@
+	Logger       logrus.FieldLogger
+	staleTimeout time.Duration
+	taskDecoder  TaskDecoder
+	scheduler    *Scheduler
+	id           string
+	dir          string
+	lastPushTime atomic.Pointer[time.Time]
+	closed       atomic.Bool
+	chunkSize    uint64
+
+	// m protects the disk operations
+	m                       sync.RWMutex
+	w                       *bufio.Writer
+	f                       *os.File
+	partialChunkSize        uint64
+	partialChunkRecordCount uint64
+	recordCount             uint64
+	chunkList               []string
+	chunkRead               map[string]struct{}
+	cursor                  int
+}
+
+type DiskQueueOptions struct {
+	// Required
+	ID          string
+	Scheduler   *Scheduler
+	Dir         string
+	TaskDecoder TaskDecoder
+
+	// Optional
+	Logger       logrus.FieldLogger
+	StaleTimeout time.Duration
+	ChunkSize    uint64
+}
+
+func NewDiskQueue(opt DiskQueueOptions) (*DiskQueue, error) {
+	if opt.ID == "" {
+		return nil, errors.New("id is required")
+	}
+	if opt.Scheduler == nil {
+		return nil, errors.New("scheduler is required")
+	}
+	if opt.Dir == "" {
+		return nil, errors.New("dir is required")
+	}
+	if opt.TaskDecoder == nil {
+		return nil, errors.New("task decoder is required")
+	}
+
+	if opt.Logger == nil {
+		opt.Logger = logrus.New()
+	}
+	opt.Logger = opt.Logger.WithField("queue_id", opt.ID)
+	if opt.StaleTimeout <= 0 {
+		opt.StaleTimeout = 5 * time.Second
+	}
+	if opt.ChunkSize <= 0 {
+		opt.ChunkSize = defaultChunkSize
+	}
+
+	q := DiskQueue{
+		id:           opt.ID,
+		scheduler:    opt.Scheduler,
+		dir:          opt.Dir,
+		Logger:       opt.Logger,
+		chunkSize:    opt.ChunkSize,
+		staleTimeout: opt.StaleTimeout,
+		taskDecoder:  opt.TaskDecoder,
+		w:            bufio.NewWriterSize(nil, 64*1024),
+		chunkRead:    make(map[string]struct{}),
+	}
+
+	// create the directory if it doesn't exist
+	err := os.MkdirAll(q.dir, 0755)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create directory")
+	}
+
+	// determine the number of records stored on disk
+	q.recordCount, err = q.calculateRecordCount()
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure the partial chunk is created or loaded
+	err = q.ensureChunk()
+	if err != nil {
+		return nil, err
+	}
+
+	// set the last push time to now
+	now := time.Now()
+	q.lastPushTime.Store(&now)
+
+	return &q, nil
+}
+
+// Close the queue, prevent further pushes and unregister it from the scheduler.
+func (q *DiskQueue) Close() error {
+	if q.closed.Swap(true) {
+		return errors.New("queue already closed")
+	}
+
+	q.scheduler.UnregisterQueue(q.id)
+
+	err := q.Flush()
+	if err != nil {
+		return err
+	}
+
+	q.m.Lock()
+	defer q.m.Unlock()
+
+	if q.f != nil {
+		err := q.f.Close()
+		if err != nil {
+			return errors.Wrap(err, "failed to close chunk")
+		}
+
+		q.f = nil
+	}
+
+	return nil
+}
+
+func (q *DiskQueue) ID() string {
+	return q.id
+}
+
+func (q *DiskQueue) Push(record []byte) error {
+	if q.closed.Load() {
```

Comment:
I think a lock may be preferred instead, otherwise the queue could be closed right after this validation passed

Reply:
That's a good point